Add docs for Web Testing Library (#1204)
testing-library/testing-library-docs: docs/web-testing-library/api.mdx

user action, timeout, or Promise. (See the
[guide to testing disappearance](guide-disappearance.mdx).)

The async methods return Promises, so be sure to use `await` or `.then` when
calling them.

## `waitFor`

```typescript
function waitFor<T>(
  callback: () => T | Promise<T>,
  options?: {
    timeout?: number
    interval?: number
    onTimeout?: (error: Error) => Error
  },
): Promise<T>
```

When in need to wait for any period of time you can use `waitFor`, to wait for
your expectations to pass. Here's a simple example:

```javascript
// ...
// Wait until the callback does not throw an error. In this case, that means
// it'll wait until the mock function has been called once.
await waitFor(() => expect(mockAPI).toHaveBeenCalledTimes(1))
// ...
```

`waitFor` may run the callback a number of times until the timeout is reached.
Note that the number of calls is constrained by the `timeout` and `interval`
options.

This can be useful if you have a unit test that mocks API calls and you need to
wait for your mock promises to all resolve.

If you return a promise in the `waitFor` callback (either explicitly or

> implicitly with `async` syntax), then the `waitFor` utility will not call your

Reviewer: ```suggestion
implicitly with the `async` syntax), then the `waitFor` utility does not call your
```

Author: I think this was copy&pasted so we want to backport these changes.

callback again until that promise rejects. This allows you to `waitFor` things
that must be checked asynchronously.

> The default `interval` is `50ms`. However it will run your callback immediately

Reviewer: ```suggestion
The default `interval` is `50ms`. However, it runs your callback immediately
```

before starting the intervals.

The default `timeout` is `1000ms`.

The default `onTimeout` takes the error and appends the `container`'s printed
state to the error message which should hopefully make it easier to track down

> what caused the timeout.

Reviewer: Isn't this specific for the DOM's implementation?

> and appends the `container`'s printed
state to the error message which should hopefully make it easier to track down
what caused the timeout.

---
id: api
title: API
---

Several utilities are provided for dealing with asynchronous code. These can be
useful to wait for an element to appear or disappear in response to an event,
user action, timeout, or Promise. (See the

> [guide to testing disappearance](guide-disappearance.mdx).)

Author: Does this link work?

Reviewer: This works, it redirects to https://deploy-preview-1204--testing-library.netlify.app/docs/guide-disappearance

  options?: {
    interval?: number
    onTimeout?: (error: Error) => Error
    showOriginalStackTrace?: boolean
    signal?: AbortSignal
    timeout?: number
  },
): Promise<T>
```

When in need to wait for any period of time you can use `waitFor`, to wait for
your expectations to pass. Here's a simple example:

```javascript
// ...
// Wait until the callback does not throw an error. In this case, that means
// it'll wait until the mock function has been called once.
await waitFor(() => expect(mockAPI).toHaveBeenCalledTimes(1))
// ...
```

`waitFor` may run the callback a number of times until the timeout is reached.
Note that the number of calls is constrained by the `timeout` and `interval`
options.

This can be useful if you have a unit test that mocks API calls and you need to
wait for your mock promises to all resolve.

If you return a promise in the `waitFor` callback (either explicitly or
implicitly with the `async` syntax), then the `waitFor` utility does not call
your callback again until that promise rejects. This allows you to `waitFor`
things that must be checked asynchronously.

The default `interval` is `50ms`. However, it runs your callback

> immediatelybefore starting the intervals.

Reviewer: ```suggestion
immediately before starting the intervals.
```
